File: src/frontend/apps/lti_site/apps/markdown/utils/tests/factories.ts
import * as faker from 'faker';

import {
  MarkdownDocument,
  MarkdownDocumentTranslation,
} from 'apps/markdown/types/models';
import { playlistMockFactory } from 'utils/tests/factories';

export const markdownDocumentMockFactory = (
  markdownDocument: Partial<MarkdownDocument> = {},
): MarkdownDocument => {
  return {
    id: faker.datatype.uuid(),
    playlist: playlistMockFactory(),
    lti_url: faker.internet.url(),
    is_draft: false,
    rendering_options: {},
    translations: [markdownTranslationMockFactory()],
    ...markdownDocument,
  };
};

export const markdownTranslationMockFactory = (
  documentTranslation: Partial<MarkdownDocumentTranslation> = {},
): MarkdownDocumentTranslation => {
  return {
    language_code: 'en',
    title: faker.lorem.word(),
    content: faker.lorem.sentence(),
    rendered_content: `<p>${faker.lorem.sentence()}</p>`,
    ...documentTranslation,
  };
};

File: src/backend/marsha/markdown/factories.py
"""Factories for the ``markdown`` app of the Marsha project."""

from django.conf import settings
from django.utils import lorem_ipsum

import factory
from factory.django import DjangoModelFactory

from marsha.core.factories import PlaylistFactory
from marsha.markdown import models


# usage of format function here is wanted in this file
# pylint: disable=consider-using-f-string
class MarkdownDocumenTranslationFactory(DjangoModelFactory):
    """Factory for the Markdown document's translation model."""

    class Meta:  # noqa
        # pylint: disable-next=protected-access
        model = models.MarkdownDocument._parler_meta.root.model

    master = factory.SubFactory(
        "marsha.markdown.factories.MarkdownDocumentFactory", translations=None
    )

    # Only translate in one language for test purpose
    language_code = settings.LANGUAGES[0][0]  # en
    title = factory.Sequence("Markdown Document's title {:03d}".format)
    content = lorem_ipsum.sentence()
    rendered_content = lorem_ipsum.sentence()


class MarkdownDocumentFactory(DjangoModelFactory):
    """Factory for the Markdown document model."""

    class Meta:  # noqa
        model = models.MarkdownDocument
        skip_postgeneration_save = True

    lti_id = factory.Faker("uuid4")
    playlist = factory.SubFactory(PlaylistFactory)

    translations = factory.RelatedFactory(
        MarkdownDocumenTranslationFactory,
        factory_related_name="master",
    )


class MarkdownImageFactory(DjangoModelFactory):
    """Factory for the MarkdownImage model."""

    markdown_document = factory.SubFactory(MarkdownDocumentFactory)
    extension = factory.fuzzy.FuzzyChoice([None, "jpg", "jpeg", "png", "gif"])

    class Meta:  # noqa
        model = models.MarkdownImage
